src/mlpack/methods/ann/layer/inception_layer.hpp

```diff
@@ -0,0 +1,290 @@
+/**
+ * @file inception_layer.hpp
+ * @author Nilay Jain
+ *
+ * Definition of the InceptionLayer class.
+ */
+
+#ifndef MLPACK_METHODS_ANN_LAYER_INCEPTION_LAYER_HPP
+#define MLPACK_METHODS_ANN_LAYER_INCEPTION_LAYER_HPP
+
+#include <mlpack/core.hpp>
+#include <mlpack/methods/ann/activation_functions/rectifier_function.hpp>
+#include <mlpack/methods/ann/layer/layer_traits.hpp>
+#include <mlpack/methods/ann/layer/one_hot_layer.hpp>
+#include <mlpack/methods/ann/layer/conv_layer.hpp>
+#include <mlpack/methods/ann/layer/pooling_layer.hpp>
+#include <mlpack/methods/ann/pooling_rules/max_pooling.hpp>
+#include <mlpack/methods/ann/layer/softmax_layer.hpp>
+#include <mlpack/methods/ann/layer/bias_layer.hpp>
+#include <mlpack/methods/ann/layer/linear_layer.hpp>
+#include <mlpack/methods/ann/layer/base_layer.hpp>
+#include <mlpack/methods/ann/network_util.hpp>
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+
+//! input --> inception_layer --> output.
+
+/*
+
+Inception module of GoogLeNet.
+
+It applies four different functions to the input array and concatenates
+their outputs along the channel dimension. Three of them are 2D
+convolutions of sizes 1x1, 3x3 and 5x5. Convolution paths of 3x3 and 5x5
+sizes have 1x1 convolutions (called projections) ahead of them. The other
+path consists of 1x1 convolution (projection) and 3x3 max pooling.
+
+The output array has the same spatial size as the input. In order to
+satisfy this, Inception module uses appropriate padding for each
+convolution and pooling.
+
+See: `Going Deeper with Convolutions <http://arxiv.org/abs/1409.4842>`_.
+
+*/
+
+template<typename InputDataType = arma::cube,
+         typename OutputDataType = arma::cube>
+class InceptionLayer
+{
+ public:
+  //! Locally-stored number of input maps.
+  size_t inMaps;
```

Comment:
When writing tests I access the layers from the InceptionLayer object. I will put the variables as private and let layers remain as public.